solver/vertex.go

```diff
@@ -0,0 +1,92 @@
+package solver
+
+import (
+	"sync"
+	"time"
+
+	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/util/progress"
+	digest "github.com/opencontainers/go-digest"
+	"golang.org/x/net/context"
+)
+
+type Vertex interface {
+	Digest() digest.Digest
+	Sys() interface{}
+	Inputs() []Input
+	Name() string // change this to general metadata
+}
+
+type Input struct {
+	Index  int
+	Vertex Vertex
+}
+
+type input struct {
+	index  int
+	vertex *vertex
+}
+
+type vertex struct {
+	mu           sync.Mutex
+	sys          interface{}
```

Comment:
This is the protobuf operation value in the case of llb. It is passed to the resolve function to find the implementation(actually should pass just `Vertex` , I'll change that). Could also connect an op implementation(eg `newExecOp()`) directly to vertex while loading, but decided to keep loading and solving separate operations.

`Reference` is a smaller interface of the `cache.ImmutableRef` as solver doesn't need to know what this value means as long as all the op implementations can accept references from other each other.